### ISSUE TITLE
Confusing way of setting nixpkgs options with lib.homeManagerConfiguration [Nix Flakes]

### ISSUE THREAD

**Issue by funketh** [nix-community/home-manager#2954]
The `nixpkgs.config` options are overridden here:
https://github.com/nix-community/home-manager/blob/5ac84ebeef5d2566ea458b81375cc6eafa19c225/flake.nix#L47-L56
This was very confusing and unintuitive since I did not know about the pkgs parameter (it isn't mentioned in the online documentation) and my `nixpkgs.config.allowUnfree = true;` in the actual configuration got ignored.

Personally, I would be against overriding these here.
(The NixOS equivalent `lib.nixosSystem` does not do this either afaik.)
At the very least the pkgs option (and its relevance for config options such as allowUnfree) should be mentioned in the docs.

**Comment by cmm** [comment]
see also #2942 
FTR I still totally don't get how `nixpkgs.config.allowUnfree = true` worked before and (subsequently) how exactly did https://github.com/NixOS/nixpkgs/commit/1c49b81263858c69b932da05ae63a7767b308e74 break it

**Comment by PaulGrandperrin** [comment]
I agree this interface is very confusing, but I would like `pkgs` to continue to be settable, otherwise `home-manager` becomes uncomposable.
In fact, I think it should be the normal way to set it, and stop using `follows` in the inputs.

Here's an example of `flake.nix`:
```nix
{
  inputs = {
    nixos-unstable.url = "github:nixos/nixpkgs/nixos-unstable";
    nixos-unstable-small.url = "github:nixos/nixpkgs/nixos-unstable-small";
    nixpkgs-darwin-unstable.url = "github:nixos/nixpkgs/nixpkgs-darwin-unstable";
 
	home-manager.url = "github:nix-community/home-manager";
    home-manager.inputs.nixpkgs.follows = "nixpkgs"; # using this method prevents composition and is confusing
  };

  outputs = inputs:
      homeConfigurations = {
	        aaa = inputs.home-manager.lib.homeManagerConfiguration {
	          pkgs = inputs.nixos-unstable.legacyPackages.x86_64-linux;
			  ...
	        };
		    bbb = inputs.home-manager.lib.homeManagerConfiguration {
	          pkgs = inputs.nixos-unstable-small.legacyPackages.x86_64-linux;
			  ...
	        };
			ccc = inputs.home-manager.lib.homeManagerConfiguration {
	          pkgs = inputs.nixpkgs-darwin-unstable.legacyPackages.x86_64-linux;
			  ...
	        };
		};
    }
}
```

otherwise, we would have to duplicate `home-manager` and hard code all of its versions in the inputs:
```nix
{
  inputs = {
    nixos-unstable.url = "github:nixos/nixpkgs/nixos-unstable";
    nixos-unstable-small.url = "github:nixos/nixpkgs/nixos-unstable-small";
    nixpkgs-darwin-unstable.url = "github:nixos/nixpkgs/nixpkgs-darwin-unstable";
 
	home-manager.url = "github:nix-community/home-manager";
    home-manager.inputs.nixpkgs.follows = "nixos-unstable";

	home-manager-small.url = "github:nix-community/home-manager";
    home-manager-small.inputs.nixpkgs.follows = "nixos-unstable-small";

	home-manager-darwin.url = "github:nix-community/home-manager";
    home-manager-darwin.inputs.nixpkgs.follows = "nixos-darwin-unstable";
  };

  outputs = inputs:
      homeConfigurations = {
	        aaa = inputs.home-manager.lib.homeManagerConfiguration {
			  ...
	        };
		    bbb = inputs.home-manager-small.lib.homeManagerConfiguration {
			  ...
	        };
			ccc = inputs.home-manager-darwin.lib.homeManagerConfiguration {
			  ...
	        };
		};
    }
```

**Comment by cmm** [comment]
FWIW I see nothing particularly wrong with this way of specifying `pkgs`, or with the fact that it is a parameter.  It's the same for NixOS, actually, the default is different but works out to superficially the same thing.

But only _superficially_ -- it seems that lately (at least) Nixpkgs acquired some bug that prevents modular overlays and nixpkgs options from working, at least when you use flakes and the `pkgs` that is provided to either `homeConfigurations` or `nixosConfigurations` is `nixpkgs.legacyPackages.${system}` or a variation (`import nixpkgs { inherit system; }` causes same misbehavior).

Basically it seems that something is subtly wrong with some magical Nixpkgs fixpoint.  I have no idea how to debug this, alas :(

**Comment by LunNova** [comment]
In #2720 I wanted to simplify this so it's easier to pass in a pkgs argument that gets used without being re-imported and configured again, but it wasn't really backwards compatible. :/

Currently I use:

```nix
        makeUser = username:
          import "${home-manager}/modules" {
            inherit pkgs;
            check = true;
            extraSpecialArgs = {
              inherit pkgs-stable;
              nixosConfig = null;
              lun = args.self;
            };
            configuration = {
              _module.args.pkgs = lib.mkForce pkgs;
              _module.args.pkgs_i686 = lib.mkForce { };
              imports = [ "${./users}/${username}" ];
              home.homeDirectory = "/home/${username}";
              home.username = "${username}";
            };
          };
```

and mkForce args.pkgs to be my own pkgs input, instead of letting HM import it again. If #2720 gets merged the same approach can be simplified by passing `useNixpkgsModule = false;` and dropping the `_module.args.pkgs*` parts.

**Comment by acowley** [comment]
I apologize if I'm not interpreting this Issue correctly, but can someone suggest how I should be setting `allowUnfree` now? I've been using this as the core of a configuration I use on NixOS, Ubuntu, and macOS machines,

```nix
homeManager.lib.homeManagerConfiguration {
  configuration = { pkgs, lib, ... }: {
    imports = [ ./common.nix ] ++ extraImports;
    nixpkgs = {
      overlays = ...;
      config = { allowUnfree = true; };
    };
  };
  inherit system homeDirectory;
  username = "acowley";
  stateVersion = "21.11";
}
```

But `allowUnfree` is not being set. Where can I set it so that it sticks?

**Comment by sandangel** [comment]
Just have this issue and overcome by overriding the `pkgs` instead

```nix
  outputs = { self, nixpkgs, nixpkgs-unstable, home-manager, ... }@inputs:
  let
    # define system, username etc...
    pkgs = import nixpkgs-unstable {
      inherit system;
      config.allowUnfree = true;
      config.allowUnsupportedSystem = true;
      overlays = [
        inputs.neovim-nightly-overlay.overlay
      ];
    };
  in {
    homeConfigurations.${username} = home-manager.lib.homeManagerConfiguration {
      inherit system username pkgs;
      configuration = import ./users/${username}/home.nix;
      homeDirectory = "/home/${username}";
      stateVersion = "22.05";
    };
}

```

**Comment by PaulGrandperrin** [comment]
As I described in #4571, `pkgs` is not actually used directly.

We can use this to our advantage and trick `home-manager` into honoring `nixpkgs.config`, `nixpkgs.system` and `nixpkgs.overlays`.

It's really neat to be able to set them in a modular way, exactly as in NixOS.

Here's an example:
```nix
homeConfigurations.test = inputs.home-manager-master.lib.homeManagerConfiguration {
	pkgs = { # it's fake but it's not this version of `pkgs` that will actually be used
		path = inputs.nixos-unstable; # path or flake to the version of `nixpkgs.pkgs` you want
		lib = path.lib; 
		overlays = []; # the arg needs to exist but we can fill the overlays in the module
		
		# you can set the system here as below, or in your module, just like with NixOS 
		#stdenv.hostPlatform.system = "x86_64-linux";
	};
	modules = [
		{
			nixpkgs = {
				system = "x86_64-linux";
				config.allowUnfree = true;
				overlays = [final: prev: { ... }];
			};
			...
		}
	];
}
```

**Comment by Mikilio** [comment]
Not stale! I have another instance of this issue as the following terminal shows:
```cmd
nix-repl> homeConfigurations."mikilio@elitebook".pkgs.config.permittedInsecurePackages
[ "electron.*" ]

nix-repl> homeConfigurations."mikilio@elitebook".config.nixpkgs.config.permittedInsecurePackages
[ "electron.*" ]

nix-repl> :q

 ❯ nix flake lock --update-input home-manager
warning: Git tree '/home/mikilio/dev/dotfiles' is dirty

 ❯ home-manager --flake .#mikilio@elitebook switch
warning: Git tree '/home/mikilio/dev/dotfiles' is dirty
error:
       … while calling the 'derivationStrict' builtin

         at /builtin/derivation.nix:9:12: (source not available)

       … while evaluating derivation 'home-manager-generation'
         whose name attribute is located at /nix/store/7lyw6yii9kja45ycj0wn74gy54wb3i91-source/pkgs/stdenv/generic/make-derivation.nix:331:7

       … while evaluating attribute 'buildCommand' of derivation 'home-manager-generation'

         at /nix/store/7lyw6yii9kja45ycj0wn74gy54wb3i91-source/pkgs/build-support/trivial-builders/default.nix:68:16:

           67|         enableParallelBuilding = true;
           68|         inherit buildCommand name;
             |                ^
           69|         passAsFile = [ "buildCommand" ]

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: Package ‘electron-28.3.3’ in /nix/store/7lyw6yii9kja45ycj0wn74gy54wb3i91-source/pkgs/development/tools/electron/binary/generic.nix:36 is marked as insecure, refusing to evaluate.


       Known issues:
        - Electron version 28.3.3 is EOL

       You can install it anyway by allowing this package, using the
       following methods:

       a) To temporarily allow all insecure packages, you can use an environment
          variable for a single invocation of the nix tools:

            $ export NIXPKGS_ALLOW_INSECURE=1

          Note: When using `nix shell`, `nix build`, `nix develop`, etc with a flake,
                then pass `--impure` in order to allow use of environment variables.

       b) for `nixos-rebuild` you can add ‘electron’ to
          `nixpkgs.config.permittedInsecurePackages` in the configuration.nix,
          like so:

            {
              nixpkgs.config.permittedInsecurePackages = [
                "electron"
              ];
            }

       c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
          ‘electron’ to `permittedInsecurePackages` in
          ~/.config/nixpkgs/config.nix, like so:

            {
              permittedInsecurePackages = [
                "electron"
              ];
            }
```

**Comment by montchr** [comment]
I've just found that overlays are not applied when added from within a configuration -- I think that might be caused by this issue.  Example using `emacs-overlay`, which provides a package called `emacs-pgtk`:

https://github.com/nix-community/emacs-overlay/blob/443981528ce117ec3055d0fd182793d6cacbe778/overlays/emacs.nix#L178

Example:

```nix
{ flake, pkgs, ... }:
{
  nixpkgs.overlays = [ flake.inputs.emacs-overlay.overlays.default ];
  # SUCCESS:
  programs.emacs.package = flake.perSystem.inputs'.emacs-overlay.packages.emacs-pgtk;
  # ERROR: no such package is available:
  programs.emacs.package = pkgs.emacs-pgtk;
}
```

**Comment by Mikilio** [comment]
Something that I have to add is that I only encountered this issue on standalone home-manager. If I use it as NixOS module, everything works.
